Closing the block inserter decrements block type impressions

dcalhoun opened the pull request:
* **Gutenberg:** Self
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3802

## Description
Fixes #33907. Closing the block inserter will decrement the block type impression counts. This results in blocks that are marked as "new" to slowly lose the designation as a user utilizes the block inserter over time. 

## How has this been tested?
1. Clone this branch. 
2. Apply the below diff. 
	<details><summary>Patch Enabling New Badge</summary>

	```diff
	diff --git a/packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js b/packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js
	index 2fcd6dc1c9..7bd08a2594 100644
	--- a/packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js
	+++ b/packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js
	@@ -19,7 +19,7 @@ function useBlockTypeImpressions( blockTypes ) {
	
				return {
					blockTypeImpressions: impressions,
	-				enableEditorOnboarding: editorOnboarding,
	+				enableEditorOnboarding: true,
				};
			},
			[]
	```

	</details>
3. Start the server and build the Demo app to an emulator/device. 
4. Open the block inserter, note the new "sparkles" badge on the Search, Embed, and Audio block buttons. 
5. Insert one of the aforementioned blocks. 
6. Open the block inserter, note the new "sparkles" badge should only be present on two of the aforementioned blocks _not_ inserted. 
7. Close the block inserter. 
8. Repeat steps 6-7 9 more times. 
9. ℹ️ **Expected:** the new "sparkles" badge is _no longer present_ on any block buttons. 

## Screenshots <!-- if applicable -->
<img width="285" alt="new-block-sparkles-badge" src="https://user-images.githubusercontent.com/438664/128391944-51323d2e-bf29-46b3-9c92-8fbb338079d8.png">

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
Closing the block inserter decrements block type impressions

In order to avoid new block badges from remaining too long, interacting
with the block inserter multiple times will cause them to eventually
disappear.

dcalhoun committed:
Increase initial new block type impression count

A higher value makes more sense now that closing the block inserter
decrements the impression count by 1.

dcalhoun committed:
Fix test by improving requestBlockTypeImpressions mock accuracy

This particular function is a React Native bridge method, which executes
a callback to send back the requested data.

mchowning commented on packages/block-library/src/index.native.js:
```diff
@@ -288,7 +288,7 @@ export const registerCoreBlocks = () => {
-	[ embed.name ]: 3,
-	[ search.name ]: 3,
-	[ audio.name ]: 3,
+	[ embed.name ]: 10,
```
10 feels a bit low to me. I'd lean toward something in the 20-40 range myself. No strong feelings though, just a thought and certainly not a blocker.

dcalhoun replied:
After reconsidering, I definitely agree. The block inserter likely is a high-touch UI element. I have increased the default impression count to 40. Addressed in https://github.com/WordPress/gutenberg/pull/33906/commits/5bf9abfe84777565e64f43822e9378f376456580.

mchowning commented on packages/block-editor/src/components/inserter/index.native.js:
```diff
@@ -161,7 +162,30 @@ export class Inserter extends Component {
-		const { onToggle } = this.props;
+		const { blockTypeImpressions, onToggle, updateSettings } = this.props;
+
+		const impressionsRemain = Object.values( blockTypeImpressions ).some(
```
What would you think about moving the `impressionsRemain` checking inside of the `!isOpen` check so we only do that work if the inserter is open. The performance gain is probably pretty minimal, so I don't feel strongly.

dcalhoun replied:
I think that is a solid improvement. Addressed by https://github.com/WordPress/gutenberg/pull/33906/commits/d187cf1696e89253ddcbc8e63b7e53d5ed8aa0af.

mchowning reviewed (approved):
Code looks good and this is working well in my testing. Left a couple of comments, but no blockers. Thanks David!

dcalhoun committed:
Refactor conditional logic to avoid computation

Iterating over the block type impressions is unnecessary when the block
inserter is opening.

dcalhoun committed:
Increase default block type impression count from 10 to 40

Interacting with the block inserter is a more frequent task, so it makes
sense to use a larger default impression count.
